pkg/ccl/backupccl/backup_job.go

```diff
@@ -230,8 +231,22 @@ func backup(
 					TotalEntryCounts:  backupManifest.EntryCounts,
 					RevisionStartTime: backupManifest.RevisionStartTime,
 				})
+
+				// Always write checkpoints to the progress folder, but only write
+				// to base directory if we detect that the cluster isn't running
+				// on pure 22.1. This is because older nodes expect checkpoints
+				// in the base directory and will throw an error if it isn't.
+				// TODO (Darryl): Figure out if this is the right way to check for pure 22.1
```

Comment:
Main thing I'm not sure about. Seems to work for if it's just a local node running 22.1 vs not 22.1, but I'm not really sure if this checks for mixed clusters or really how to test for that.

Reply:
The canonical way to ensure that all nodes are running on a single version is to introduce a new cluster version in `cockroach_versions.go` and then check that the CV `IsActive`. I would recommend adding a `BackupDoesNotOverwriteLatestAndCheckpoint` version and checking on that.

As a high-level comment, I think it is important to lay out all the mixed version states we can get into for backup, restore and show backup. In particular, interacting with an older backup from a 22.1 node.

1) For the LATEST file:

- A backup chain that was started on 21.2 or older will always write to `LATEST`. For this reason, I think we want a backup started on 22.1 to attempt to read from both the directory and just the file itself. If we only read from the directory, it is possible that the incremental on the 22.1 node does not find any latest file, and this would cause `BACKUP INTO LATEST IN...` to fail? Similarly, we could be using `RESTORE` or `SHOW BACKUP` from a 22.1 node on an older backup. All these codepaths should be first checking if the backup has a latest dir, and if not falling back to reading just the LATEST file.

2) For the checkpoint file:

- What if a 21.2 node started the backup and wrote a checkpoint file. Then a 22.1 node resumed the backup. If it only reads the directory, it won't find anything. I think we should be reading from the directory, and if that fails check for the existence of the top level file.

- On the flip side, what if a 22.1 node started the backup, and a 21.2 node resumed it. I think what you're doing here correctly deals with that case.

Without being too prescriptive, what I'd like are two helpers used to write and read the latest directory/file (and similarly two helpers for the checkpoint file). We can then push the version checks, and "first try dir, if that fails try file" logic into those two helpers.